Interpolated strings

Nigel-Ecma opened the pull request:
…e a first draft.

Once the decision on how much of the ANLTR language we would use was made the grammar
was fairly straghtforward.

The prose describing the feature is more of a challenge as it is a language-level
surfacing of features of the Standard Library, and those features are either defined
in another Standard or not defined in any Standard but which we pretend are…

So the challenge is how much of the semantics should be included somewhere in this
Standard: what is a `FormattableString` vs. a `string` for the result type of an
*interpolated_string_expression*; what is the field width, what is the format, etc.?

This draft punts on some of these (e.g. the first one), includes details (e.g. for the the second
one), and even a new *informative* appendix, §C.4 (for the third one).

Is this balance right?

Well that depends on how we view our aim here…

Nigel-Ecma committed:
Initial draft for interpolated strings; the grammar is good, the prose a first draft.

Once the decision on how much of the ANLTR language we would use was made the grammar
was fairly straghtforward.

The prose describing the feature is more of a challenge as it is a language-level
surfacing of features of the Standard Library, and those features are either defined
in another Standard or not defined in any Standard but which we pretend are…

So the chanllenge is how much of the semantics should be included somewhere in this
Standard: what is a `FormattableString` vs. a `string` for the result type of an
*interpolated_string_expression*; what is the field width, what is the format, etc.?

This draft punts on some of these (e.g. the first one), includes details (e.g. for the the second
one), and even a new *informative* appendix, §C.4 (for the third one).

Is this balance right?

Well that depends on how we view our aim here…

jskeet commented on standard/expressions.md:
```diff
@@ -1150,6 +1151,200 @@ object o = (new int[3])[1];
+    : ':' Interpolated_Verbatim_String_Element+
+    ;
+
+Interpolated_Verbatim_String_End
```
Do we particularly need this to be separate from Interpolated_Regular_String_End? Could we just have Interpolated_String_End and use it in both rules? Or is that harder due to the context-sensitive nature?

Same applies for some other rules as well, I believe - it's just that this was the simplest one to validate.

Nigel-Ecma replied:
Correct, it is the context-sensitivity.

In the working ANTLR grammar there are two (in addition to the standard one) modes, one for each of regular & verbatim interpolated strings. In these two modes some rules are different while others are the same but with different names (as there is a single scope for names). This is a common situation when using modes, nothing special for the use case here.

jskeet replied:
SGTM.

jskeet commented on standard/expressions.md:
```diff
@@ -1150,6 +1151,200 @@ object o = (new int[3])[1];
+    : constant_expression
+    ;
+
+Interpolated_Regular_String_Start
```
I can't remember the different casing rules we've got, and I probably couldn't check whether these are appropriate anyway. I strongly suspect they're all fine, but I bring it up so that someone with more of a clue can review that aspect specifically.

Nigel-Ecma replied:
I suspect it looks a bit off as this is the first place where we really intermingle lexer & parser rules a lot. The casing is correct or ANTLR would have barfed.

jskeet commented on standard/expressions.md:
```diff
@@ -1150,6 +1151,200 @@ object o = (new int[3])[1];
+other tokens in the language. *end note*
+
+*Note:* The above grammar is not ANTLR-ready due to the context sensitive lexical rules. As with
+other lexer generators ANTLR supports context sensitive lexical rule, for example using its *lexical modes*,
```
lexical rule => lexical rules

(unless I'm misreading)

Nigel-Ecma replied:
Thanks, will be fixed.

jskeet commented on standard/expressions.md:
```diff
@@ -1150,6 +1151,200 @@ object o = (new int[3])[1];
+
+*Note:* The differences between the possible types an *interpolated_string_expression* may be determined from the documentation for `System.String` ([§C.2](standard-library.md#c2-standard-library-types-defined-in-isoiec-23271)) and `System.FormattableString` ([§C.3](standard-library.md#c3-standard-library-types-not-defined-in-isoiec-23271)). *end note*
+
+The meaning on an interpolation, both *regular_interpolation* and *verbatim_interpolation*, is to format the value of the *expression* as a `string` either according to the format specified by the *Regular_Interpolation_Format* or *Verbatim_Interpolation_Format*, or according to a default format for the type of *expression*. The formatted string is then modified by the *interpolation_minimum_widt ...
```
meaning on => meaning of

Nigel-Ecma replied:
Thanks, will be fixed.

jskeet commented on standard/expressions.md:
```diff
@@ -1150,6 +1151,200 @@ object o = (new int[3])[1];
+
+*Note:* How the default format for a type is determined is detailed in the documentation for `System.String` ([§C.2](standard-library.md#c2-standard-library-types-defined-in-isoiec-23271)) and `System.FormattableString` ([§C.3](standard-library.md#c3-standard-library-types-not-defined-in-isoiec-23271)). Descriptions of standard formats, which are identical for *Regular_Interpolation_Format* and * ...
+
+In an *interpolation_minimum_width* the *constant_expression* shall have an implicit conversion to `int` and it termed the *alignment*, the *field width* is the absolute value of the alignment:
```
it termed => is termed

I'd suggest a semi-colon before "the *field width*" rather than a comma... or possibly even a period, with a new sentence at the end of the paragraph of something like: "The alignment and field width are used to determine the padding, if any, to be used when formatting the string:"

Nigel-Ecma replied:
Will fix the typo.

The prose here is not good. The source material often uses "alignment" when they mean "field width" and generally mixes the two up. I've partially fixed this but it is still wrong – the alignment is the *sign* of the *constant_expression* (not its value), the field width is its absolute value. More wordsmithing required...

BillWagner replied:
Proposal: If the only remaining issue is with some of the prose, let's merge this and open a new issue to update the prose in this section.

jskeet replied:
Nigel will tidy it up a little, but not worry too much about achieving perfection: we can adjust slightly later.

jskeet commented on standard/expressions.md:
```diff
@@ -1150,6 +1151,200 @@ object o = (new int[3])[1];
+| :----------------------------------- | :------------------------------------------------------------ | :----------- |
+| `$"{text}"`                          | `string.Format("{0}", text)`                                  | `"red"`      |
+| `$"{{text}}"`                        | `string.Format("{{text}})`                                    | `"{text}"`   |
+| `$"{ text , 4 }"`                    | `string.Format("{0,4}", text)`                                | `" red"`     |
```
Is it worth having an example that shows the interpolation_minimum_width being an expression using a named constant rather than just digits?

jskeet replied:
Example of valid code that surprised me:

```csharp
const int Width = 4;
string text = "abc";
string formatted = $"{text, Width}";
```

jskeet commented on standard/standard-library.md:
```diff
@@ -658,3 +652,400 @@ namespace System.Runtime.CompilerServices
+it is converted to a string. Either standard or custom formats can be used. A
+standard format takes the form *Axx*, where *A* is a single
+alphabetic character called the *format specifier*
+, and *xx* is an integer between zero and 99 inclusive, called the *precision specifier*. The format specifier controls the type
```
Move the comma to the previous line, here and below? (In general we might want to rewrap this paragraph.)

Nigel-Ecma replied:
See below…

jskeet replied:
Resolving as we are going to keep just the XML as-is.

jskeet replied:
Unresolving due to an extra space before the comma being incorrect.

jskeet commented on standard/standard-library.md:
```diff
@@ -658,3 +652,400 @@ namespace System.Runtime.CompilerServices
+or more of the standard format specifiers.]
+
+[*Note:* When a format includes symbols that vary by culture, such as the currency
+symbol included by the ‘C’ and ‘c’ formats, a formatting object supplies the
```
Do we want to use curly quotes here?

Nigel-Ecma replied:
Keep going…

jskeet replied:
Resolving not to change.

jskeet commented on standard/standard-library.md:
```diff
@@ -658,3 +652,400 @@ namespace System.Runtime.CompilerServices
+alphabetic character called the *format specifier*
+, and *xx* is an integer between zero and 99 inclusive, called the *precision specifier*. The format specifier controls the type
+of formatting applied to the value being represented as a string. The
+*precision specifier* controls the number 
```
Precision specifiers are only relevant for numeric format strings, of course... there are standard formats for date/time values too.

jskeet replied:
(I appreciate this is a copy/paste job, and we may well not want to bother changing it at all.)

Nigel-Ecma replied:
I too pondered over how much to change this. The original style and formatting doesn't even fit the rest of the CLI Standard, but then it is distributed as an XML file that was mechanically produced from CLR documentation so it wasn't really noticable. I mainly did a mechanical translation into markdown and then a few hand tweaks so it's still recognisable as the source material.

This is the largest single block of format specs I spotted but it is about numeric types, the closing para directs the reader to `Enum` and `DateTime`. How much to include?

I think TG2 needs to ask whether it should foster (not adopt) the whole of the standard library docs while TG3 is in hibernation, to be returned when TG3 wakes up. That would of course address the "how much" issue, and would allow TG2 to remove §C.3 as it would be empty and TG2 would no longer have to fudge and pretend these the types in §C.3 have been Standardised…

BillWagner replied:
Hmm. I'd be concerned unless we can expand the membership in TG2, including some people who have more expertise in this area.

jskeet commented on standard/standard-library.md:
```diff
@@ -658,3 +652,400 @@ namespace System.Runtime.CompilerServices
+object members that are used with numeric data types in the Base Class
+Library.
+
+<table>
```
I seem to remember that I make very specific assumptions about the formats of tables in the Word converter. This may very well not work at the moment.

Nigel-Ecma replied:
I think it looks like the other tables, and the Word Convertor didn't barf but I've not seen its output…

jskeet replied:
Jon to attach a screenshot after fetching and running the Word converter.

jskeet replied:
Nope, the Word converter fails entirely for this. This will take a fair amount of work... I'll see if the same table converter that manages to handle the table in the unsafe types section would work here - but it does mean a rather larger tweak of the XML we're "copying" the content from.

jskeet commented on standard/standard-library.md:
```diff
@@ -658,3 +652,400 @@ namespace System.Runtime.CompilerServices
+digits are copied to the output string immediately before the first Zero or
+Digit placeholder.
+
+A custom format can contain
```
Again, this is all about numeric format strings.

Nigel-Ecma replied:
Look two comments back

jskeet reviewed (commented):
Various comments - but above all, thanks for your work on this.

I'm not at all sure about C.4 at the moment. It doesn't sit quite right, but I don't know the best solution to that. I need to ponder, then look forward to a discussion in the meeting next week.

BillWagner commented on standard/expressions.md:
```diff
@@ -1150,6 +1151,200 @@ object o = (new int[3])[1];
+
+An *interpolated_string_expression* is classified as a value. If it is immediately converted to `System.IFormattable` or `System.FormattableString` with an implicit interpolated string conversion (§implicit-interpolated-string-conversions), the interpolated string expression has that type. Otherwise, it has the type `string`.
+
+*Note:* The differences between the possible types an *interpolated_string_expression* may be determined from the documentation for `System.String` ([§C.2](standard-library.md#c2-standard-library-types-defined-in-isoiec-23271)) and `System.FormattableString` ([§C.3](standard-library.md#c3-standard-library-types-not-defined-in-isoiec-23271)). *end note*
```
More a note for Nigel:

If you just write "§C.2" and "§C.3", the section numbering tool will convert that text to the correct link.

BillWagner replied:
Note formatting:

```suggestion
> *Note:* The differences between the possible types an *interpolated_string_expression* may be determined from the documentation for `System.String` ([§C.2](standard-library.md#c2-standard-library-types-defined-in-isoiec-23271)) and `System.FormattableString` ([§C.3](standard-library.md#c3-standard-library-types-not-defined-in-isoiec-23271)). *end note*
```

Nigel-Ecma replied:
Fixed

BillWagner commented on standard/expressions.md:
```diff
@@ -1150,6 +1151,200 @@ object o = (new int[3])[1];
+| *Interpolated_Regular_String_End* | Only recognised after an *Interpolated_Regular_String_Start* and only if any intervening tokens are either *Interpolated_Regular_String_Mid*s or tokens that can be part of *regular_interpolation*s, including tokens for any *interpolated_regular_string_expression*s contained within such interpolations. |
+| *Interpolated_Verbatim_String_Mid* *Verbatim_Interpolation_Format* *Interpolated_Verbatim_String_End* | Recognition of these three rules follows that of the corresponding rules above with each mentioned *regular* grammar rule replaced by the corresponding *verbatim* one. |
+
+*Note:* The above rules are context sensitive as their definitions overlap with those of
+other tokens in the language. *end note*
```
All our other "Note"s have been indented:

```suggestion
> *Note:* The above rules are context sensitive as their definitions overlap with those of other tokens in the language. *end note*
```

Nigel-Ecma replied:
Fixed

BillWagner commented on standard/expressions.md:
```diff
@@ -1150,6 +1151,200 @@ object o = (new int[3])[1];
+*Note:* The above rules are context sensitive as their definitions overlap with those of
+other tokens in the language. *end note*
+
+*Note:* The above grammar is not ANTLR-ready due to the context sensitive lexical rules. As with
+other lexer generators ANTLR supports context sensitive lexical rule, for example using its *lexical modes*,
+but this is an implementation detail and therefore not part of this Standard. *end note*
```
Same comment (and made the change Jon suggested):

```suggestion
> *Note:* The above grammar is not ANTLR-ready due to the context sensitive lexical rules. As with other lexer generators ANTLR supports context sensitive lexical rules, for example using its *lexical modes*, but this is an implementation detail and therefore not part of this Standard. *end note*
```

Nigel-Ecma replied:
Fixed

BillWagner commented on standard/expressions.md:
```diff
@@ -1150,6 +1151,200 @@ object o = (new int[3])[1];
+
+The meaning on an interpolation, both *regular_interpolation* and *verbatim_interpolation*, is to format the value of the *expression* as a `string` either according to the format specified by the *Regular_Interpolation_Format* or *Verbatim_Interpolation_Format*, or according to a default format for the type of *expression*. The formatted string is then modified by the *interpolation_minimum_widt ...
+
+*Note:* How the default format for a type is determined is detailed in the documentation for `System.String` ([§C.2](standard-library.md#c2-standard-library-types-defined-in-isoiec-23271)) and `System.FormattableString` ([§C.3](standard-library.md#c3-standard-library-types-not-defined-in-isoiec-23271)). Descriptions of standard formats, which are identical for *Regular_Interpolation_Format* and * ...
```
same:

```suggestion
> *Note:* How the default format for a type is determined is detailed in the documentation for `System.String` ([§C.2](standard-library.md#c2-standard-library-types-defined-in-isoiec-23271)) and `System.FormattableString` ([§C.3](standard-library.md#c3-standard-library-types-not-defined-in-isoiec-23271)). Descriptions of standard formats, which are identical for *Regular_Interpolation_Format* and *Verbatim_Interpolation_Format*, may be found in the documentation for `System.IFormattable` ([§C.4](standard-library.md#c4-format-specifications)) and in other types in the standard library ([§C](standard-library.md#annex-c-standard-library)). *end note*
```

Nigel-Ecma replied:
Fixed

BillWagner commented on standard/expressions.md:
```diff
@@ -1150,6 +1151,200 @@ object o = (new int[3])[1];
 
+### §interpolated-string-expressions Interpolated string expressions
+
+An *interpolated_string_expression* consists of a `$` character immediately followed by text within `"` characters. Within the quoted text there are zero or more **_interpolations_** delimited by `{` and `}` characters, each of which encloses an *expression* and optional formatting specifications.
```
Q for @jskeet 

Will the markdown converter parse the `**_` for bold italic, or should we use `***`?

jskeet replied:
Good question - no idea; I believe `***` would work though and is what we've done elsewhere.

BillWagner commented on standard/expressions.md:
```diff
@@ -1150,6 +1151,200 @@ object o = (new int[3])[1];
+
+Interpolated string expressions have two forms; regular (*interpolated_regular_string_expression*)
+and verbatim (*interpolated_verbatim_string_expression*); which are lexically similar to, but differ semantically from, the two forms of string
+literals ([§7.4.5.6](lexical-structure.md#7456-string-literals)).
```
Note for @Nigel-Ecma 

You can just write the section number for the link, and the tool will fill in the correct section:

```suggestion
literals (§7.4.5.6).
```

Nigel-Ecma replied:
OK, changed. Does it make any difference as the long hand version will be there next time, or is it just to help reduce errors?

BillWagner replied:
It won't make any difference later. It's to save you time when you add links.

BillWagner commented on standard/expressions.md:
```diff
@@ -1150,6 +1151,200 @@ object o = (new int[3])[1];
+| Rule | Contextual Requirements |
+| :--- | :---------------------- |
+| *Interpolated_Regular_String_Mid* | Only recognised after an *Interpolated_Regular_String_Start*, between any following interpolations, and before the corresponding *Interpolated_Regular_String_End*. |
+| *Regular_Interpolation_Format* | Only recognised immediately before the closing delimiter (`}`) of a *regular_interpolation*. |
```
Should this key on the `:` in the interpolation rather than "Only recognized immediately before the closing `}` delimiter?  Would this rule mean that a conditional (`? :`) expression should parse correctly without parentheses if there's an additional format specifier after the expression?

Nigel-Ecma replied:
The tricky part here is the format can contain just about anything, including parentheses etc. So the rule matches if there is a colon, some text, and then a following '}' (which is not part of the token itself) *and* the whole token is at the outmost level.

What we're not making clear here, because its an implementation thing, is that in order to implement the context sensitivity you need to track *all* bracket kinds.

The question is whether the text here is a correct/sufficient description so that implementors can produce a correct implementation. I'll elaborate on the TG2 list…

BillWagner replied:
I like the alternative you suggested to the committee.  Adding it here for discussion:

```suggestion
| *Regular_Interpolation_Format* | Only recognised within a *regular_interpolation*, when the starting colon (`:`) is not nested within any kind of bracket (parentheses/braces/square), and when the following brace (`}`) is the terminating delimiter of a *regular_interpolation*. |
```

jskeet replied:
Agreed with the suggestion without `, and when the following brace (`}`) is the terminating delimiter of a *regular_interpolation*`.

BillWagner reviewed (approved):
Overall, I like this. I had a few smaller comments, but none that are required.

I'm ready to approve it.

jskeet commented on standard/standard-library.md:
```diff
@@ -658,3 +652,400 @@ namespace System.Runtime.CompilerServices
+following table describes the characters that are used in constructing custom
+formats.
+
+<table>
```
Another table for Jon to screenshot

jskeet commented:
Resolution from meeting of 2021-11-17:

- Nigel to check in new version after applying suggestions, typo fixes etc as per comments
- Nigel to change from draft to "ready to review"
- Jon to check all changes decided in meeting have been applied
- Jon to run Word converter and take screenshots of long comment in grammar, and tables in the appendix C.4
- Mads to review and confirm he's happy (as he wasn't at the meeting)
- Nigel to press the button and merge :)

Nigel-Ecma committed:
- Fixed the various typos people found
- Fixed some gaffs in the XML -> MD translation in §C.4
- Extended the examples to include a defined constant and left-alignment
- Normalised bold italic as *** and not **_ or _**
- Improved context sensitive condition for Regular_Interpolation_Format
- Snuck in the removal of a spurious *end note* from an earlier PR

jskeet commented on standard/expressions.md:
```diff
@@ -1150,6 +1151,205 @@ object o = (new int[3])[1];
+    | Close_Brace_Escape_Sequence
+    ;
+
+fragment Interpolated_Regular_String_Character
```
This didn't come out well in the Word converter :(

![Screenshot (15)](https://user-images.githubusercontent.com/17011/142574873-28effd54-1f6c-4b8a-af96-24547e414371.png)

I'll think about options here. Given that it's only a single rule, how would you feel about putting the comments before the production?

jskeet commented on standard/standard-library.md:
```diff
@@ -658,3 +652,393 @@ namespace System.Runtime.CompilerServices
+alphabetic character called the *format specifier*, and *xx* is an integer between zero and 99 inclusive, called the *precision specifier*. The format specifier controls the type
+of formatting applied to the value being represented as a string. The
+*precision specifier* controls the number 
+of significant digits or decimal places in the string, if applicable. [*Note:* For the list of standard format specifiers, see the
```
This note needs to be changed to use the format we use elsewhere. (Currently the converter thinks it's a link.) I'd suggest making it a new paragraph - I'm not sure how "inline note which isn't in a list" works at the moment.

jskeet commented on standard/standard-library.md:
```diff
@@ -658,3 +652,393 @@ namespace System.Runtime.CompilerServices
+table below. Note that a given data type, such as `System.Int32`, might not support one
+or more of the standard format specifiers.]
+
+[*Note:* When a format includes symbols that vary by culture, such as the currency
```
Ditto (note format)

BillWagner commented on standard/standard-library.md:
```diff
@@ -658,3 +652,393 @@ namespace System.Runtime.CompilerServices
+alphabetic character called the *format specifier*, and *xx* is an integer between zero and 99 inclusive, called the *precision specifier*. The format specifier controls the type
+of formatting applied to the value being represented as a string. The
+*precision specifier* controls the number 
+of significant digits or decimal places in the string, if applicable. [*Note:* For the list of standard format specifiers, see the
+table below. Note that a given data type, such as `System.Int32`, might not support one
+or more of the standard format specifiers.]
```
Suggestion for formatting:

```suggestion
of significant digits or decimal places in the string, if applicable. 

> *Note:* For the list of standard format specifiers, see the table below. Note that a given data type, such as `System.Int32`, might not support one or more of the standard format specifiers. *end note*
```

BillWagner commented on standard/standard-library.md:
```diff
@@ -658,3 +652,393 @@ namespace System.Runtime.CompilerServices
+table below. Note that a given data type, such as `System.Int32`, might not support one
+or more of the standard format specifiers.]
+
+[*Note:* When a format includes symbols that vary by culture, such as the currency
+symbol included by the ‘C’ and ‘c’ formats, a formatting object supplies the
+actual characters used in the string representation. A method might include a
+parameter to pass a `System.IFormatProvider` object that supplies a
+formatting object, or the method might use the default formatting object, which
+contains the symbol definitions for the current culture. The current culture
+typically uses the same set of symbols used system-wide by default. In the Base
+Class Library, the formatting object for system-supplied numeric types is a
+`System.Globalization.NumberFormatInfo` instance. For `System.DateTime` instances, a 
+`System.Globalization.DateTimeFormatInfo` is
+used.]
```
Suggestion to use the standard *note* format:

```suggestion
> *Note:* When a format includes symbols that vary by culture, such as the currencysymbol included by the ‘C’ and ‘c’ formats, a formatting object supplies the actual characters used in the string representation. A method might include a parameter to pass a `System.IFormatProvider` object that supplies a formatting object, or the method might use the default formatting object, which contains the symbol definitions for the current culture. The current culture typically uses the same set of symbols used system-wide by default. In the Base Class Library, the formatting object for system-supplied numeric types is a `System.Globalization.NumberFormatInfo` instance. For `System.DateTime` instances, a `System.Globalization.DateTimeFormatInfo` is used. *end note*
```

BillWagner reviewed (approved):
Once we update the note formats, this is ready.

BillWagner committed:
Apply suggestions from code review
